core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java

```diff
@@ -127,58 +124,66 @@ private Object handleSingleKeyWriteCommand(InvocationContext ctx, DataWriteComma
       return invokeNextThenApply(ctx, command, handleSingleKeyWriteReturn);
    }
 
-   private Object handleSingleKeyWriteReturn(InvocationContext ctx, DataWriteCommand dataWriteCommand, Object rv) throws Throwable {
-      int segment = dataWriteCommand.getSegment();
-      if (dataWriteCommand.isSuccessful() &&
-            clusteringDependentLogic.getCacheTopology().getSegmentDistribution(segment).isPrimary()) {
-         CacheEntry entry = ctx.lookupEntry(dataWriteCommand.getKey());
-         DataWriteCommand crossSiteCommand;
-         if (entry.isRemoved()) {
-            crossSiteCommand = commandsFactory.buildRemoveCommand(dataWriteCommand.getKey(), null, segment,
-                  dataWriteCommand.getFlagsBitSet());
-         } else {
-            crossSiteCommand = commandsFactory.buildPutKeyValueCommand(dataWriteCommand.getKey(), entry.getValue(),
-                  segment, entry.getMetadata(), dataWriteCommand.getFlagsBitSet());
+   private Object handleSingleKeyWriteReturn(InvocationContext ctx, DataWriteCommand dataWriteCommand, Object rv) {
+      if (!dataWriteCommand.isSuccessful()) {
+         if (trace) {
+            log.tracef("Command %s is not successful, not replicating", dataWriteCommand);
          }
-         InvocationStage stage = backupSender.backupWrite(crossSiteCommand, dataWriteCommand);
-         return stage.thenReturn(ctx, dataWriteCommand, rv);
+         return rv;
+      }
+      int segment = dataWriteCommand.getSegment();
+      DistributionInfo dInfo = clusteringDependentLogic.getCacheTopology().getSegmentDistribution(segment);
+      if (dInfo.isWriteOwner()) {
+         //all owners has to keep track of the updates even if the primary owner is the only one who sends it.
+         iracManager.trackUpdatedKey(dataWriteCommand.getKey(), dataWriteCommand.getCommandInvocationId());
+      }
+      if (dInfo.isPrimary()) { //primary owner sends for sync backups
+         CacheEntry<?,?> entry = ctx.lookupEntry(dataWriteCommand.getKey());
+         WriteCommand crossSiteCommand = createCommandForXSite(entry, segment, dataWriteCommand.getFlagsBitSet());
+         return backupSender.backupWrite(crossSiteCommand, dataWriteCommand).thenReturn(ctx, dataWriteCommand, rv);
```

Comment:
Sorry I can't quite follow, but what this is for? Isn't the irac manager handling the replication of values and the prior if stores the keys in backup owners as well, I guess miss what part this does.

Maybe `hasEnabledBackups` should be `hasSyncEnabledBackups` ? at EmptyConstructorNamedCacheFactory

```java
         return configuration.sites().hasEnabledBackups() ?
                new BackupSenderImpl() :
                NoOpBackupSender.getInstance();
```

Reply:
so, IRAC handles the async xsite backups and the BackupSender handles only the sync xsite backups. 
I'll change the EmptyConstructorNamedCacheFactory

Reply:
Okay, that is what I kinda though, so we just need to make sure that we only use the BackupSender if we have a sync backup then.

Reply:
already changed! :+1: